protobuf/protoc-gen-gogrpcmock/plugin/plugin.go

```diff
@@ -0,0 +1,277 @@
+)
+
+var (
+	rxID          = regexp.MustCompile("(?i)id")
+	rxEmail       = regexp.MustCompile("(?i)email")
+	rxPhone       = regexp.MustCompile("(?i)phone")
+	rxURL         = regexp.MustCompile("(?i)url")
+	rxDescription = regexp.MustCompile("(?i)^(description|desc)$")
+	rxLatitude    = regexp.MustCompile("(?i)^(latitude|lat)$")
+	rxLongitude   = regexp.MustCompile("(?i)^(longitude|lng)$")
+)
+
+type grpcmock struct {
+	*generator.Generator
+	contextPkg string
+}
+
+func New() generator.Plugin {
+	return &grpcmock{}
+}
+
+func (g *grpcmock) Name() string {
+	return "grpcmock"
+}
+
+func (g *grpcmock) Init(gen *generator.Generator) {
+	g.Generator = gen
+}
+
+func (g *grpcmock) Generate(file *generator.FileDescriptor) {
+	if len(file.FileDescriptorProto.Service) == 0 {
+		return
+	}
+
+	for _, service := range file.FileDescriptorProto.Service {
+		g.mockService(file, service)
+	}
+}
+
+func (g *grpcmock) GenerateImports(file *generator.FileDescriptor) {
+	if len(file.FileDescriptorProto.Service) == 0 {
+		return
+	}
+	imports := generator.NewPluginImports(g.Generator)
+	g.contextPkg = imports.NewImport("context").Use()
+	imports.GenerateImports(file)
+}
+
+func (g *grpcmock) mockService(file *generator.FileDescriptor, service *descriptor.ServiceDescriptorProto) {
+	origServName := service.GetName()
+	servName := generator.CamelCase(origServName)
+	servTypeName := fmt.Sprintf("%sMock", servName)
+
+	g.P(`type `, servTypeName, ` struct {}`)
+	g.P()
+	for _, method := range service.Method {
+		g.mockMethod(servTypeName, method)
+	}
+}
+
+func (g *grpcmock) mockMethod(servTypeName string, method *descriptor.MethodDescriptorProto) {
+	if method.GetServerStreaming() || method.GetClientStreaming() {
+		// TODO: [RC] Mock streaming methods.
+		return
+	}
+
+	methName := generator.CamelCase(method.GetName())
+	inType := g.typeName(method.GetInputType())
+	outType := g.typeName(method.GetOutputType())
+
+	g.P(`func (m *`, servTypeName, `) `, methName, `(ctx context.Context, req *`, inType, `) (*`, outType, `, error){`)
+	g.In()
+
+	msg := g.objectNamed(method.GetOutputType())
+	if m, ok := msg.(*generator.Descriptor); ok && !m.GetOptions().GetMapEntry() {
+		g.P(`res := `)
+		// TODO: Depth should be a cmd param
+		g.generateMockMessage(m, false, true, 5)
+		g.P(`return res, nil`)
+	} else {
+		// Should this return an error?
+		g.P(`return nil, nil`)
+	}
+
+	g.Out()
+	g.P(`}`)
+}
+
+func (g *grpcmock) generateMockMessage(msg *generator.Descriptor, inner, nullable bool, depth int) {
+	depth--
+	if depth < 0 {
+		return
+	}
+
+	msgName := g.TypeName(msg)
+	if nullable {
+		msgName = "&" + msgName
+	}
+
+	g.P(msgName, `{`)
+	g.In()
+
+	for _, field := range msg.Field {
+		fieldName := g.GetFieldName(msg, field)
+		fieldType, _ := g.GoType(msg, field)
+		repeated := field.IsRepeated()
+		nullable := gogoproto.IsNullable(field) && field.IsMessage()
+
+		if field.OneofIndex != nil {
+			// TODO: Deal with oneof
```

Comment:
Should we be adding an INFO log here and at other places to say we are skipping some fields?

Reply:
Will be addressing these soon; I just wanted to get a first pass working.